### PR TITLE
Perf improvements

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142431

A series of directed perf improvements to drive down the dynamo tracing cost of the given test. Before this PR the compile took about 60s, and after takes 30s. Individual improvements are listed below along with the approximate improvement of that change.

Tested with this model:
```
@torch.compile(backend="eager")
def model_add(x, y):
    out = x
    for i in range(5000):
        out = torch.add(out, y)
    return out
```

- import in build: 60.48 -> 59.92
    Stop importing builder in the inner loop of `VariableTracker.build()`
- import_module: 59.92 -> 52.9
    Using `importlib.import_module()` within the hot path of symbolic_convert is slow. Memoize it.
- cache on import_source: 52.9 -> 52.58
    add cache to `InstructionTranslatorBase.import_source()`
- direct Guard: 52.58 -> 51.76
    Remove explicit constant check from `VariableBuilder.install_guards()`
the args calling convention.  Also remove a lambda binding.
- Guard slots: 51.76 -> 51.34
    Add slots to Guard
- c++ VariableTrackerCache: 51.34 -> 50.36
    Move VariableTrackerCache into C++. 
- kill import: 50.36 -> 49.12
    Stop importing in the body of `BuiltinVariable.call_getattr()`
- c strip_function_call: 49.12 -> 47.77
    Convert `strip_function_call()` into C++
- slight improvement on __instancecheck__: 47.77 -> 47.62
    Switch out an `isinstance()` for an `is` in the very hot `VariableTrackerMeta.__instancecheck__`.
- no import on hot path: 47.62 -> 47.26
    Removed another `import` in the body of the hot path.
- don't unnecessarily call getframeinfo on the hot path: 47.26 -> 37.66
    Stop using `getframeinfo()` when we only care about the function name and throw the rest away.
- don't call expensive _set_guard_export_info if it's a duplicate guard: 37.66 -> 34.86
    Move the call to `_set_guard_export_info()` after the duplicate guard check in `GuardBuilder.DUPLICATE_INPUT()`
- bytecode_transform improvements: 34.86 -> 33.9
    Use slots on InstructionExnTabEntry and Instruction.
    Stop doing python version checks in the middle of `convert_instruction()` and `inst_has_op_bits()`.
- no re in arg_ref: 33.9 -> 33.7
    Avoid use of python re and move valid varname check in `GuardBuilder.arg_ref()` into C++
- cache cleaned_instructions: 33.7 -> 30.0
    Cache the interesting/expensive bits of `cleaned_instructions()`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by aorenste** [comment]
This is the HUD for this change: https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Wed%2C%2004%20Dec%202024%2001%3A30%3A03%20GMT&stopTime=Wed%2C%2011%20Dec%202024%2001%3A30%3A03%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/aorenste/153/head&lCommit=634e707e994d6e3ed4ee6e24b65695b420a2d737&rBranch=gh/aorenste/153/base&rCommit=8bf28b3613b3fc13ff5cc69295b4c9adae7a7313

**Comment by ezyang** [comment]
@laithsakka I was hoping to see an improvement in pr time benchmarks here, but I do not see it. What happened?

**Comment by ezyang** [comment]
This is looooong. No chance of getting stacked?

**Comment by aorenste** [comment]
> This is looooong. No chance of getting stacked?

I actually have it stacked locally but was told by someone else to make it a single PR... lol.

File: torch/csrc/dynamo/variable_tracker_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+#include <functional>
+#include <stdexcept>
+
+namespace {
+
+namespace pyptr = torch::pyptr;
+
+struct VariableTrackerCacheKey {
+  uintptr_t m_value;
```
Comment (jansel) [inline torch/csrc/dynamo/variable_tracker_cache.cpp:12]: Add some comments to the new C++ sources to explain how things works. It is not totally clear from first read that this corresponds to the `id()` (memory address) of a Python object.

File: torch/csrc/dynamo/variable_tracker_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  void clear() {
+    m_cache.clear();
+  }
+
+  pyptr::Owned clone() {
+    throw std::runtime_error("unimplemented VariableTrackerCache::clone");
+  }
+
+  pyptr::Owned lookup(pyptr::Borrowed value, pyptr::Borrowed source) const {
+    auto it = m_cache.find(VariableTrackerCacheKey(value, source));
```
Comment (jansel) [inline torch/csrc/dynamo/variable_tracker_cache.cpp:87]: Does constructing this increment the reference count to source?  Possibly that could be avoided.

File: torch/utils/_functools.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    cache_name = "_cache_method_" + f.__name__
+
+    @functools.wraps(f)
+    def wrap(self: _C, *args: _P.args, **kwargs: _P.kwargs) -> _T:
+        assert not kwargs
+        if not (cache := getattr(self, cache_name, None)):
+            cache = {}
+            setattr(self, cache_name, cache)
+        if args in cache:
+            return cache[args]
```
Comment (jansel) [inline torch/utils/_functools.py:35]: Could avoid two cache lookups here with `.get()`+sentinal or try/except.

File: torch/csrc/utils/python_ptr.h [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  Owned to_owned() const;
+
+ protected:
+  PyObject* m_value;
+};
+
+/// Represents an owned PyObject*. You must be explicit on acquire (either with
+/// Owned::own() or Owned::steal()). Releases the
+/// reference on release.
+struct Owned : Borrowed {
```
Comment (jansel) [inline torch/csrc/utils/python_ptr.h:62]: We have some similar classes already in PyTorch.  For example `THPObjectPtr`.  There are also the pybind classes that do the same thing and are used in PyTorch.  PyBind is slow for doing actually binding and C++/Python conversions, but I think the object/handle stuff should be fine.

Can we reuse the existing things?

**Review by jansel — CHANGES_REQUESTED** [review]
I agree that smaller stacked PRs are easier to review.  Left some comments below though.

File: torch/csrc/utils/python_ptr.h (hunk 4 above)
Comment (aorenste) [reply torch/csrc/utils/python_ptr.h:62]: I originally wrote this using pybind. Doing it this way was measurably faster. I looked around for something that already existed but didn't find THPObjectPtr (hard to figure out what to search for). I'll switch to using that.

**Comment by aorenste** [comment]
Replaced with the stack ending here: https://github.com/pytorch/pytorch/pull/143070

**Comment by laithsakka** [comment]
> @laithsakka I was hoping to see an improvement in pr time benchmarks here, but I do not see it. What happened?

there is 
```
please update all results that changed significantly, and not only the failed ones
PASS: benchmark ('add_loop_eager_dynamic', 'compile_time_instruction_count') pass, actual result 5635755069 -1.18% is within expected 5703000000 ±2.50%

WIN: benchmark ('add_loop_inductor', 'compile_time_instruction_count') failed, actual result 29014572914 is -1.68% lower than expected 29510000000 ±1.50% please update the expected results. 

please update all results that changed significantly, and not only the failed ones
PASS: benchmark ('add_loop_inductor_dynamic_gpu', 'compile_time_instruction_count') pass, actual result 43104370761 -0.41% is within expected 43280000000 ±2.50%

PASS: benchmark ('add_loop_inductor_gpu', 'compile_time_instruction_count') pass, actual result 25525701[723](https://github.com/pytorch/pytorch/actions/runs/12264841956/job/34226068054#step:22:724) -0.64% is within expected 25690000000 ±1.50%

WIN: benchmark ('basic_modules_ListOfLinears_eager', 'compile_time_instruction_count') failed, actual result 910371103 is -11.87% lower than expected 1033000000 ±1.50% please update the expected results. 

please update all results that changed significantly, and not only the failed ones
PASS: benchmark ('basic_modules_ListOfLinears_inductor', 'compile_time_instruction_count') pass, actual result 20535380258 -1.32% is within expected 20810000000 ±1.50%

PASS: benchmark ('basic_modules_ListOfLinears_inductor_gpu_force_shape_pad', 'compile_time_instruction_count') pass, actual result 16825567734 -1.14% is within expected 17020000000 ±1.50%

PASS: benchmark ('basic_modules_ListOfLinears_inductor_gpu', 'compile_time_instruction_count') pass, actual result 16921091598 -1.96% is within expected 17260000000 ±20.00%

PASS: benchmark ('update_hint_regression', 'compile_time_instruction_count') pass, actual result 1670235126 +0.07% is within expected 1669000000 ±2.00%

WIN: benchmark ('sum_floordiv_regression', 'compile_time_instruction_count') failed, actual result 1008435799 is -9.39% lower than expected 1113000000 ±1.50% please update the expected results. 

please update all results that changed significantly, and not only the failed ones
PASS: benchmark ('symint_sum', 'compile_time_instruction_count') pass, actual result 3287649241 -0.16% is within expected 3293000000 ±1.50%

WIN: benchmark ('aotdispatcher_inference_nosubclass_cpu', 'compile_time_instruction_count') failed, actual result 19[724](https://github.com/pytorch/pytorch/actions/runs/12264841956/job/34226068054#step:22:725)22156 is -2.26% lower than expected 2018000000 ±1.50% please update the expected results. 

please update all results that changed significantly, and not only the failed ones
WIN: benchmark ('aotdispatcher_inference_subclass_cpu', 'compile_time_instruction_count') failed, actual result 5672910493 is -2.91% lower than expected 5843000000 ±1.50% please update the expected results. 

please update all results that changed significantly, and not only the failed ones
PASS: benchmark ('aotdispatcher_partitioner_cpu', 'compile_time_instruction_count') pass, actual result 9053353302 -0.46% is within expected 9095000000 ±1.50%

PASS: benchmark ('aotdispatcher_training_nosubclass_cpu', 'compile_time_instruction_count') pass, actual result 3814119867 -1.27% is within expected 3863000000 ±1.50%

WIN: benchmark ('aotdispatcher_training_subclass_cpu', 'compile_time_instruction_count') failed, actual result 10201010[726](https://github.com/pytorch/pytorch/actions/runs/12264841956/job/34226068054#step:22:727) is -2.01% lower than expected 10410000000 ±1.50% please update the expected results.
```
idk why is it catgorized as unrelated.

**Pull request closed without merging**